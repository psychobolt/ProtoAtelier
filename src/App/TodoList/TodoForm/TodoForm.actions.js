let nextTodoId = 0;

export const Actions = {
  ADD_TODO: 'addTodo',
};

export const addTodo = text => ({
  type: Actions.ADD_TODO,
  todo: {
    id: nextTodoId++, // eslint-disable-line no-plusplus
    text,
  },
});
