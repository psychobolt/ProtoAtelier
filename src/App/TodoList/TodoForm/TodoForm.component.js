// @flow
import React from 'react';
// @flow
import { XInput, type XInputEvent } from 'Framework/ReactXelToolkit';
import styles from './TodoForm.style';

type Props = {
  inputValue: string,
  onValueSubmit: (value: string) => void,
  style: {},
};

type State = {
  inputValue: string,
}

export const KEYCODE_ENTER = 13;

export default class TodoForm extends React.Component<Props, State> {
  static defaultProps = {
    inputValue: '',
    style: {},
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      inputValue: props.inputValue,
    };
  }

  onButtonClick = () => {
    if (this.state.inputValue) {
      this.submit();
    }
  }

  onInputKeyup = (event: XInputEvent) => {
    if (event.keyCode === KEYCODE_ENTER) {
      this.submit();
    } else {
      this.setState({
        inputValue: event.target.value,
      });
    }
  }

  submit() {
    this.props.onValueSubmit(this.state.inputValue);
    this.setState({
      inputValue: '',
    });
  }

  props: Props;

  render() {
    return (
      <x-box style={this.props.style}>
        <XInput
          style={styles.input}
          value={this.state.inputValue}
          onKeyup={this.onInputKeyup}
        />
        <x-button onClick={this.onButtonClick}>
          <x-label>Add Todo</x-label>
        </x-button>
      </x-box>
    );
  }
}
