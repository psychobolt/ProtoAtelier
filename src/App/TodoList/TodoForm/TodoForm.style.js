export default {
  input: {
    marginRight: '6px',
  },
};
