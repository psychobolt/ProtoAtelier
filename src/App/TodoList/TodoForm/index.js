export * from './TodoForm.actions';
export { default } from './TodoForm.container';
