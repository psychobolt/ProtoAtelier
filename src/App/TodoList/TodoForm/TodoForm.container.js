// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';

import TodoForm from './TodoForm.component';
import { addTodo } from './TodoForm.actions';

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  onValueSubmit: (value) => {
    if (value.trim()) {
      dispatch(addTodo(value));
    }
  },
});

export default connect(undefined, mapDispatchToProps)(TodoForm);
