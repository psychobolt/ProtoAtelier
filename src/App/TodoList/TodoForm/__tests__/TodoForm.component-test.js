import React from 'react';
import { shallow, mount } from 'enzyme';

import TodoForm, { KEYCODE_ENTER } from '../TodoForm.component';

const KEYCODE_1 = 49;

describe('component <TodoForm />', () => {
  it('should render correctly', () => {
    const props = {
      onValueSubmit: jest.fn(),
    };
    const wrapper = shallow(<TodoForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state on keyup', () => {
    const props = {
      onValueSubmit: jest.fn(),
    };
    const value = '1';
    const wrapper = mount(<TodoForm {...props} />);
    const input = wrapper.find('input');
    input.node.value = value;
    input.simulate('keyup', {
      keyCode: KEYCODE_1,
    });
    expect(props.onValueSubmit.mock.calls.length).toBe(0);
    expect(wrapper.state().inputValue).toBe(value);
  });

  it('should submit form on key ENTER', () => {
    const props = {
      onValueSubmit: jest.fn(),
    };
    const wrapper = mount(<TodoForm {...props} />);
    wrapper.find('input').simulate('keyup', {
      keyCode: KEYCODE_ENTER,
    });
    expect(props.onValueSubmit.mock.calls.length).toBe(1);
  });
});
