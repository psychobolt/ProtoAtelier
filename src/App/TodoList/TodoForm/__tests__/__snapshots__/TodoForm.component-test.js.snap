// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <TodoForm /> should render correctly 1`] = `
<x-box
  style={Object {}}
>
  <XInput
    fallback={[Function]}
    onKeyup={[Function]}
    style={
      Object {
        "marginRight": "6px",
      }
    }
    value=""
  />
  <x-button
    onClick={[Function]}
  >
    <x-label>
      Add Todo
    </x-label>
  </x-button>
</x-box>
`;
